frame/contracts/src/migration/v14.rs

```diff
@@ -0,0 +1,283 @@
+		#[codec(compact)]
+		pub refcount: u64,
+		pub determinism: Determinism,
+		pub code_len: u32,
+	}
+
+	#[storage_alias]
+	pub type CodeInfoOf<T: Config, OldCurrency> =
+		StorageMap<Pallet<T>, Twox64Concat, CodeHash<T>, CodeInfo<T, OldCurrency>>;
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub fn store_dummy_code<T, OldCurrency>()
+where
+	T: Config,
+	OldCurrency: ReservableCurrency<<T as frame_system::Config>::AccountId> + 'static,
+{
+	use frame_benchmarking::account;
+	use sp_runtime::traits::Hash;
+	use sp_std::vec;
+
+	let len = T::MaxCodeLen::get();
+	let code = vec![42u8; len as usize];
+	let hash = T::Hashing::hash(&code);
+
+	let info = old::CodeInfo {
+		owner: account::<T::AccountId>("account", 0, 0),
+		deposit: u32::MAX.into(),
+		refcount: u64::MAX,
+		determinism: Determinism::Enforced,
+		code_len: len,
+	};
+	old::CodeInfoOf::<T, OldCurrency>::insert(hash, info);
+}
+
+#[cfg(feature = "try-runtime")]
+#[derive(Encode, Decode)]
+/// Accounts for the balance allocation of a code owner.
+struct BalanceAllocation<T, OldCurrency>
+where
+	T: Config,
+	OldCurrency: ReservableCurrency<<T as frame_system::Config>::AccountId>,
+{
+	/// Total reserved balance as storage deposit for the owner.
+	reserved: old::BalanceOf<T, OldCurrency>,
+	/// Total balance of the owner.
+	total: old::BalanceOf<T, OldCurrency>,
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, DefaultNoBound)]
+pub struct Migration<T, OldCurrency>
+where
+	T: Config,
+	OldCurrency: ReservableCurrency<<T as frame_system::Config>::AccountId>,
+{
+	last_code_hash: Option<CodeHash<T>>,
+	_phantom: PhantomData<(T, OldCurrency)>,
+}
+
+impl<T, OldCurrency> MigrationStep for Migration<T, OldCurrency>
+where
+	T: Config,
+	OldCurrency: 'static + ReservableCurrency<<T as frame_system::Config>::AccountId>,
+{
+	const VERSION: u16 = 14;
+
+	fn max_step_weight() -> Weight {
+		T::WeightInfo::v14_migration_step()
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_hash) = self.last_code_hash.take() {
+			old::CodeInfoOf::<T, OldCurrency>::iter_from(
+				old::CodeInfoOf::<T, OldCurrency>::hashed_key_for(last_hash),
+			)
+		} else {
+			old::CodeInfoOf::<T, OldCurrency>::iter()
+		};
+
+		if let Some((hash, code_info)) = iter.next() {
+			log::debug!(target: LOG_TARGET, "Migrating storage deposit for 0x{:?}", HexDisplay::from(&code_info.owner.encode()));
+
+			let remaining = OldCurrency::unreserve(&code_info.owner, code_info.deposit);
+
+			if remaining > Zero::zero() {
+				log::warn!(
+					target: LOG_TARGET,
+					"Code owner's account 0x{:?} for code {:?} has some non-unreservable deposit {:?} from a total of {:?} that will remain in reserved.",
+					HexDisplay::from(&code_info.owner.encode()),
+					hash,
+					remaining,
+					code_info.deposit
+				);
+			}
+
+			let unreserved = code_info.deposit.saturating_sub(remaining);
+
+			let amount = if let Ok(amount) = BalanceOf::<T>::decode(&mut &unreserved.encode()[..]) {
```

Comment:
Bringing @pgherveou's question back from a removed file comment.
> this seems weird, there are no more direct way to convert to the desired value?
Agreed, I gave it a second thought, and actually there should be a `From` bound and leave the details to the implementation.

Something on these lines
```rust
BalanceOf<T>: From<old::BalanceOf<T, OldCurrency>>
...
let amount = unreserved.into();
```